rust-runtime/aws-smithy-http/src/event_stream/receiver.rs

```diff
@@ -156,10 +156,19 @@ impl<T, E> Receiver<T, E> {
                     raw: RawMessage::Decoded(message),
                 }),
             },
-            Err(err) => Err(SdkError::ResponseError {
-                err: Box::new(err),
-                raw: RawMessage::Decoded(message),
-            }),
+            Err(err) => {
+                if matches!(
+                    err,
+                    aws_smithy_eventstream::error::Error::EventStreamError(_)
+                ) {
+                    // Terminate the stream
+                    self.buffer = RecvBuf::EosPartial(Default::default())
+                }
```

Comment:
I think it would be better to keep `unmarshall` immutable on `self`, and if the buffer needs to be changed, the caller of `unmarshall` should set it if it returns `Err` at all. All errors would be terminal on this. The code can be rearranged so that `unmarshall` is only called once in `recv`.

`RecvBuf::EosPartial` doesn't seem like the right variant to use here since it specifically means: some data was received, it wasn't a full message. A new variant should be created instead.